Шеломов Никита

nikshel opened the pull request.

nikshel committed:
Первая версия

Zhigalov commented on counter.js:
```diff
@@ -0,0 +1,110 @@
+    var dictionary = {};
+    for (var i = 0; i < 10; i++) {
+        ['verstka-tasks-', 'javascript-tasks-'].forEach(function (repoName) {
+            var fullRepoName = repoName + (i + 1).toString();
```
Так может итератор сразу сделать в интервале `[1..10]` а не в `[0..9]`
Он вроде больше ни для чего не нужен

nikshel replied:
Сделаю.

Zhigalov commented on counter.js:
```diff
@@ -0,0 +1,110 @@
+    if (resultJson.encoding != 'base64') {
+        throw new Error('Формат не поддерживается');
+    }
+    return new Buffer(resultJson.content, 'base64').toString('utf-8');
```
Расскажи какую проблему решают эти преобразования? Почему нельзя просто вернуть `result.getBody('utf-8')`?

nikshel replied:
У API гитхаба метод /repos/:owner/:repo/readme возвращет json'овский документ со всякой полезной инфой про ридми. В том числе в поле content лежит само содержимое файла. А так как json - текстовый формат, то байты закодированы в символы через base64

Zhigalov commented on counter.js:
```diff
@@ -0,0 +1,110 @@
+    words.forEach(function (word) {
+        var stem = natural.PorterStemmerRu.stem(word);
+        if (dictionary[stem] == undefined) {
+            dictionary[stem] = [1, word];
```
1. Почему именно массив? Кажется что объект будет выглядеть семантичнее.
2. Зачем сохраять `word`?

nikshel replied:
1. Можно и объект сделать.
2. В процессе заполнения словаря я обрезаю слово до его основы (чтобы отсечь падежи и все в таком духе). А для функции top(n) мне нужно хотя бы одно полное слово. Так что я пошел самым простым путем и просто сохранил первое встреченное слово для каждой основы.

Zhigalov replied:
1. Не честно хранить только первое слово, мне кажется. Может переформулировать вывод и говорить что с корнем/основой таким-то найдено столько-то слов. Это честнее мне кажется. Тогда второй элемент массива будет не нужен

Zhigalov commented:
Синхронный запрос это конечно чит :)
На самом деле там на каждый процесс создается дочерний и это очень дорогая операция. В продакшн коде у тебя не будет возможности так делать.

nikshel commented:
@Zhigalov 
Я так и не понял, как можно в данном случае использовать асинхронные запросы. В идеале хотелось бы так:
1. запустить кучу асинхронных запросов
2. в коллбэках парсить данные и складывать в словарь
3. в конце функции top (или count) "подождать", пока они все отработают
4. вернуть из функции результат

Проблема, собственно, в шаге №4: коллбэки будут вызваны уже после того, как завершиться код, из которого была вызвана функция top (count).

Zhigalov commented:
> коллбэки будут вызваны уже после того, как завершиться код, из которого была вызвана функция top (count).

Нужно принимать параметр-коллбек в функцию top, это будет правильно.

> в конце функции top (или count) "подождать", пока они все отработают

Посмотри в сторону [async.js](https://github.com/caolan/async)

nikshel committed:
Добавил немного асинхронности

1. теперь все работает через коллбэки
2. исправлен интервал в getDictionary()
3. изменен вывод: теперь выводится топ основ слова

nikshel commented:
@Zhigalov 
Сделал.
А вообще мне казалось, что сигнатура функций top и count была частью условия задачи, так что я даже не думал о том, чтобы добавить параметр - коллбэк

Zhigalov commented on counter.js:
```diff
@@ -0,0 +1,129 @@
+            .map(function (item) {
+                return item.stem + ' --- ' + item.count;
+            });
+        callback(result);
```
Принято принимать в колбек первым параметром ошибку (если она есть), а вторым - данные. Чтобы различать с успехом закончилась операция или нет.

Это конечно не более чем договоренность, но практически везде она работает. Советую её использовать и привыкать к ней

Zhigalov commented on counter.js:
```diff
@@ -0,0 +1,129 @@
+}
+
+function getDictionary(callback) {
+    if (cachedDictionary != undefined) {
```
Лучше `!==`

Zhigalov commented on counter.js:
```diff
@@ -0,0 +1,129 @@
+                    var words = getWordsFromText(text);
+                    addWords(dictionary, words);
+                    console.log('Обработан репозиторий  ' + fullRepoName);
+                    localCallback();
```
В качестве альтернативы: можно запараллелить вызовы `getReadme`, а `getWordsFromText` делать в колбеке который в `async.parallel`.

Почему мне кажется это лучше - меньше уровень вложенности + сейчас есть `localCallback` который не участвует в логике приложения а нужен только для синхронизации вызовов. Мне кажется от таких вещей лучше избавляться чтобы код "не шумел".

Эту правку я оставляю на твое усмотрение, она на сдачу задачи не повлияет

Zhigalov commented on counter.js:
```diff
@@ -0,0 +1,129 @@
+        }
+        var resultJson = JSON.parse(body);
+        if (resultJson.encoding != 'base64') {
+            throw new Error('Формат не поддерживается');
```
Вот тут какраз ошибка вылетит в трубу.
Лучше ее передать первым аргументом в коллбек

Zhigalov commented:
В целом все ок, нужно доработать сигнатуру коллбеков (первый аргумент ошибка, второй - данные)

nikshel committed:
Исправил коллбэки

1. в коллбэки передаются ошибки
2. сначала асинхронно отрабатывают getReadme, потом парсятся результаты

nikshel commented:
@Zhigalov 
Исправил.
Хотя я не уверен, что копить результаты всех запросов, а потом уже их парсить - это хорошая идея. Если бы тексты были большими, то они бы могли съесть много памяти.

Zhigalov commented:
Все верно, если тексты большие они не влезут в память. 
Если это действительно будет так - надо будет переписать код, заюзать стримы, вести поточную обработку.
В той постановке которая есть сейчас - такой код решает задачу хорошо. Он наглядный и простой

Задачу принял!

Zhigalov commented:
:star: что разобрался с коллбеками
